cli/src/main.rs

```diff
@@ -0,0 +1,464 @@
+    println!("  Accounts:\n");
+    for account in &instr.accounts {
+        println!(
+            "    * {}\n      signer: {}, writable: {}\n",
+            account.pubkey, account.is_signer, account.is_writable,
+        );
+    }
+
+    if instr.program_id == bpf_loader_upgradeable::ID
+        && bpf_loader_upgradeable::is_upgrade_instruction(&instr.data[..])
+    {
+        // Account meaning, according to
+        // https://docs.rs/solana-sdk/1.5.19/solana_sdk/loader_upgradeable_instruction/enum.UpgradeableLoaderInstruction.html#variant.Upgrade
+        println!("  This is a bpf_loader_upgradeable::upgrade instruction.");
+        println!("    Program to upgrade:      {}", instr.accounts[1].pubkey);
+        println!("    Program data address:    {}", instr.accounts[0].pubkey);
+        println!("    Buffer with new program: {}", instr.accounts[2].pubkey);
+        println!("    Spill address:           {}", instr.accounts[3].pubkey);
+    } else {
+        println!("  Unrecognized instruction.");
+    }
+}
+
+fn propose_upgrade(program: Program, opts: ProposeUpgradeOpts) {
+    let (program_derived_address, _nonce) =
+        get_multisig_program_address(&program, &opts.multisig_address);
+
+    let upgrade_instruction = bpf_loader_upgradeable::upgrade(
+        &opts.program_address,
+        &opts.buffer_address,
+        // The upgrade authority is the multisig-derived program address.
+        &program_derived_address,
+        &opts.spill_address,
+    );
+
+    // The program expects `multisig::TransactionAccount` instead of
+    // `solana_sdk::AccountMeta`. The types are structurally identical,
+    // but not nominally, so we need to convert these.
+    let accounts: Vec<_> = upgrade_instruction
+        .accounts
+        .iter()
+        .map(multisig::TransactionAccount::from)
+        .collect();
+
+    // The transaction is stored by the Multisig program in yet another account,
+    // that we create just for this transaction.
+    // TODO: Should we save the private key, to allow deleting the multisig
+    // account in order to recover the funds?
+    let transaction_account = Keypair::generate(&mut OsRng);
+    println!("Transaction account: {}", transaction_account.pubkey());
+
+    program
+        .request()
+        // Create the program-owned account that will hold the transaction data,
+        // and fund it from the payer account to make it rent-exempt.
+        .instruction(system_instruction::create_account(
+            &program.payer(),
+            &transaction_account.pubkey(),
+            // TODO: Is there a good way to determine the size of the
+            // transaction; can we serialize and measure maybe? For now, assume
+            // 500 bytes will be sufficient.
+            // TODO: Ask for confirmation from the user first before funding the
+            // account.
+            program
+                .rpc()
+                .get_minimum_balance_for_rent_exemption(500)
```

Comment:
Check how Anchor does serialization and set a constant for the maximum number of owners in an account so we can know its size.

Reply:
I implemented this in 06f9540cd90a0c2bd75f7a016626b2a1610beca1, as part of #3 to avoid conflicts.